### PR TITLE
sql: switch CREATE DATABASE to same RequireSuperUser helper

### PR DESCRIPTION
**Pull request by dt**
followup fix from #14009

Also fix some `err` scope assignments and update PlanHookState comment.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14028)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-Keo4YwaRo7acZgfaqh1:bric211) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
https://reviewable.io/reviews/cockroachdb/cockroach/14009#-Keo-skn6ns6mtj9pvll:

> does passing the dir actually matter here? this permission validation doesn't seem to require a backup to be written.

---

*[pkg/sql/create.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4eNOI5Qf1bUjjcIn:-Keo4eNOI5Qf1bUjjcIo:b-vgag88) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/sql/create.go#L82)):*
> ```Go
> 	}
> 
> 	if err := p.RequireSuperUser("CREATE DATABASE"); err != nil {
> ```

the other caller just passes "RENAME"; why the inconsistency?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-Keo9huJItL6D4E5tnb8:b-ypxxtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

https://reviewable.io/reviews/cockroachdb/cockroach/14009#-Keo-skn6ns6mtj9pvll:

> does passing the dir actually matter here? this permission validation doesn't seem to require a backup to be written.
</blockquote></details>

oh, then why is `into_db=system` needed? I'm sure there's something I'm missing here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-Keo9mE8xRLZRsRkuzwD:bwe0tdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

oh, then why is `into_db=system` needed? I'm sure there's something I'm missing here.
</blockquote></details>

user _is_ root but does not have CREATE on `system`.

---

*[pkg/sql/create.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4eNOI5Qf1bUjjcIn:-Keo9Ed2DY5opEmsq4pf:b-40zpyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/sql/create.go#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

the other caller just passes "RENAME"; why the inconsistency?
</blockquote></details>

`CREATE DATABASE` is the actual statement being run here, but there is no `RENAME DATABASE` statement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-KeoAz2cr245qnBYT6To:bozhx23) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

user _is_ root but does not have CREATE on `system`.
</blockquote></details>

To recap: We read descriptors from a backup, we assign them parents (using their parentIDs or `into_db`), then we check that we have CREATE on every parent to which we're adding a desc.

So we need to point it at some descriptors, and we need to point those descriptors somewhere root doesn't have CREATE, like system.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-KeoBlF6WYFwLKQj2fkN:b20fwcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

To recap: We read descriptors from a backup, we assign them parents (using their parentIDs or `into_db`), then we check that we have CREATE on every parent to which we're adding a desc.

So we need to point it at some descriptors, and we need to point those descriptors somewhere root doesn't have CREATE, like system.
</blockquote></details>

I see. Isn't it true though that if `into_db` is specified, checking the parents can be done before reading the descriptors?

---

*[pkg/sql/create.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4eNOI5Qf1bUjjcIn:-KeoC3AuswQRL7y5P0Zk:b-4fcjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/sql/create.go#L82)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

`CREATE DATABASE` is the actual statement being run here, but there is no `RENAME DATABASE` statement.
</blockquote></details>

So then shouldn't the other string be "ALTER DATABASE ... RENAME TO" or some other pointer to the statement? I don't care which way you go, but I'd like this to be consistent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-KeoF9HOJ8V2VNFY9AXZ:b-81sgjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I see. Isn't it true though that if `into_db` is specified, checking the parents can be done before reading the descriptors?
</blockquote></details>

yes, as a special-case, when `into_db` is set, we could do an early permission check, but scattering multiple, slightly different permissions checks though other logic seems less maintainable and harder to reason about than a simple, single assertion that we have privs on every assigned parent, no matter how it was assigned.

---

*[pkg/sql/create.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4eNOI5Qf1bUjjcIn:-KeoGITdNxfKk9lskNoH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/sql/create.go#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

So then shouldn't the other string be "ALTER DATABASE ... RENAME TO" or some other pointer to the statement? I don't care which way you go, but I'd like this to be consistent.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/ccl/sqlccl/backup_test.go, line 807 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-Keo4YwaRo7acZgfaqh0:-KeoGcpClTRzi1bRieWZ:bypttv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba0216fc4521abcf35630014dad33bc1d6f0117/pkg/ccl/sqlccl/backup_test.go#L807)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

yes, as a special-case, when `into_db` is set, we could do an early permission check, but scattering multiple, slightly different permissions checks though other logic seems less maintainable and harder to reason about than a simple, single assertion that we have privs on every assigned parent, no matter how it was assigned.
</blockquote></details>

OK. Consider adding this comment (I'm not demanding that you do).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-:-KeoGia0gw8uGPZmSuJd:bnfp4nl)*

**Comment by benesch** [comment]
:lgtm:

---

Review status: 6 of 8 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14028#-:-KeoPod2qXLPOufVQrNA:bnfp4nl)*
